python/ray/rllib/agents/ddpg/ddpg_policy_graph.py

```diff
@@ -499,59 +388,185 @@ def set_weights(self, weights):
 
     @override(PolicyGraph)
     def get_state(self):
-        return [TFPolicyGraph.get_state(self), self.cur_epsilon]
+        return [
+            TFPolicyGraph.get_state(self), self.cur_noise_scale,
+            self.cur_pure_exploration_phase
+        ]
 
     @override(PolicyGraph)
     def set_state(self, state):
         TFPolicyGraph.set_state(self, state[0])
         self.set_epsilon(state[1])
+        self.set_pure_exploration_phase(state[2])
 
     def _build_q_network(self, obs, obs_space, action_space, actions):
-        q_net = QNetwork(
-            ModelCatalog.get_model({
-                "obs": obs,
-                "is_training": self._get_is_training_placeholder(),
-            }, obs_space, action_space, 1, self.config["model"]), actions,
-            self.config["critic_hiddens"],
-            self.config["critic_hidden_activation"])
-        return q_net.value, q_net.model
-
-    def _build_p_network(self, obs, obs_space, action_space):
-        policy_net = PolicyNetwork(
-            ModelCatalog.get_model({
-                "obs": obs,
-                "is_training": self._get_is_training_placeholder(),
-            }, obs_space, action_space, 1, self.config["model"]),
-            self.dim_actions, self.config["actor_hiddens"],
-            self.config["actor_hidden_activation"],
-            self.config["parameter_noise"])
-        return policy_net.action_scores, policy_net.model
-
-    def _build_action_network(self, p_values, stochastic, eps,
-                              is_target=False):
-        return ActionNetwork(
-            p_values, self.low_action, self.high_action, stochastic, eps,
-            self.config["exploration_theta"], self.config["exploration_sigma"],
-            self.config["smooth_target_policy"], self.config["act_noise"],
-            is_target, self.config["target_noise"],
-            self.config["noise_clip"]).actions
+        if self.config["use_state_preprocessor"]:
+            q_model = ModelCatalog.get_model(
```

Comment:
@joneswong what do you think about the refactoring here to inline the QNetwork / PolicyNetwork classes?

Reply:
@qxcv  What's the purpose of this factoring? I have not got the advantage against original implementation. Could you give a brief explanation for me?

Reply:
To build the Q network, the action network, or the DDPG loss, `DDPGPolicyGraph.__init__` would previously have to call a `self._build_*()` method, which would then pass its arguments & some config variables onto a class that did the actual work of building the network or loss (e.g `QNetwork`/`PolicyNetwork`). I didn't see any point in having both a `self._build_*()` method _and_ a separate function or class to carry out the actual work of that method, so I removed one of those layers of indirection. This also made the code fractionally shorter and easier to debug.

As for the `use_state_preprocessor` thing: previously the code was passing the state through a model from `ModelCatalog` (by default, a two-layer FC net), _then_ passing it through an extra two network layers. That was not at all obvious to me the first time I read the code, and I suspect it was unintentional (you don't need a four-layer network to solve Mujoco problems). I decided to disable the preprocessing thing by default, but also added an option to turn it back on in case somebody wants to work with image-based domains, in which case they'll probably want to preprocess the state with a shallow convnet.